### ISSUE TITLE
Migrate from EBS account_number to org_id

### ISSUE THREAD

**Issue by infinitewarp** [cloudigrade/cloudigrade#1135]
This issue is *blocked* until we have a general timeline from sources-api. When will they actually feed us new users/etc. with the org-id? We could start work before sources-api is ready, but if we have already migrated known users and new users aren't including the org-id, we'd have to keep looking up and migrating users as they arrive. Let's try to avoid that complexity by simply waiting until sources-api is done. See sources work here: https://issues.redhat.com/browse/RHCLOUD-18531

## Summary

As a member of the Red Hat console-dot platform, I need to change my primary user identifier from the EBS account_number to the org_id.

## Acceptance Criteria

- [x] Verify cloudigrade appropriately handles auth for incoming HTTP requests that include:
  - [x] ONLY account_number, NOT org_id (current code, likely swatch requests for some time as well)
  - [x] BOTH the org_id, AND account_number (transition?)
    - When we get both, which do we actually use for auth? Do we check both? If we get both, trust the org_id.
  - [x] ONLY the org_id, NOT account_number (future?)
- [x] Verify cloudigrade includes org_id on outgoing requests to sources using:
  - [x] HTTP
  - [x] Kafka

## Assumptions and Questions

- [Slides presented on 2022-03-15 tech talk from Jozef Hartinger](https://docs.google.com/presentation/d/1LsrkKvYQ9sJr0FlNbHLV6OvLYGZbkCgLaL3stLHm-1w/edit#slide=id.g1177c88155f_0_6)
- Today, the incoming 3scale header's account_number *is* the EBS account number (for all intents and purposes).
- We expect that the platform will provide an http service for looking up and translating values for migration.
- Do we need to preserve the old account_number values for any period of time?
  - Yes, we should know both account_number and org_id for some unspecified transition period.
  - This means we should have an initial migration to fetch the org_ids for any existing users.
  - Then we need to retain both the account_number and org_id for new incoming users. The latest 3scale headers should include *both* values for the foreseeable future. Example:
    ```
    {
        "identity": {
            "type": "User"
            "account_number": "901578",
            "org_id": "5318290",
            "user": {
                "first_name": "Jozef",
                "is_active": true,
                "is_internal": true,
                "last_name": "Hartinger",
                "locale": "en_US",
                "is_org_admin": false,
                "username": "jharting",
                "email": "jharting@redhat.com"
            },
            "internal": {
                "auth_time": 0,
                "auth_type": "jwt-auth",
                "org_id": "5318290"
            },
        },
    }
    ```
  - Eventually we will phase out and drop the account_number. Date unknown.
  - This also means our incoming HTTP header auth handling should accept *either* x- header for account_number or org_id for some period.
    - We currently accept `x-rh-cloudigrade-account-number`. We should also start accepting `x-rh-cloudigrade-org-id`.
  - HTTP and Kafka requests we make to sources-api must start including an x- header for org_id.
    - We currently use `x-rh-sources-account-number` for talking to sources. We need to confirm, but we probably need to add `x-rh-sources-org-id` to requests that we make to source-api.
- When/how will the integrations with sources-api change?
  - TBD! See also: https://coreos.slack.com/archives/C0246P60U8H/p1647456215808659
  - https://issues.redhat.com/browse/RHCLOUD-18531
- When/how will the integrations with tally change?
  - TBD! Add information here when we know.
- After we have confirmed that both tally and sources-api have started accepting and using org-id-based auth, we should drop support for account_number. This work TBD in a future issue.
- Question: In the case where an incoming request has the org-id, do we auth from the org-id exclusively? Do we need to cross-check the account_number also?
  - Normally, we might say "trust but verify", but in this case, we have high confidence that the headers are originating from trustworthy actors. If we get an org-id, we just use the org-id, especially since the plan is to phase out and drop the account_number.
- In summary, the changes we need to make are:
  1. find a place to put the new org-id
  2. migrate all existing users
  3. start saving the org-id for new users
  4. start accepting and using org-id for auth in incoming HTTP requests
  5. start including org-id headers in outgoing HTTP and Kafka messages to sources

**Comment by infinitewarp** [comment]
As discussed on 2022-04-13 backlog refinement, we think the work for sources, etc. is complete (or nearing complete?) but we don't have the documentation/issue immediately available. Please add links here to relevant issues/docs when starting development.

**Comment by infinitewarp** [comment]
See also email titled `ConsoleDot migration of tenant identifier to Organization ID` sent from Josh Barnard to several mailing lists on 2022-04-12.

**Comment by abellotti** [comment]
While only a go client is available for services to call the 3scale gateway's translation API (Ebs account number to Org_Id) here https://github.com/RedHatInsights/tenant-utils, there are no plans on writing a Python equivalent. Also, while one can run a cli for conversion, for our case where we're not really using an org_id attribute in the database, it's best for us to call the api directly for conversion.  See the following:

- https://coreos.slack.com/archives/C03B7U5AE67/p1650913400288639
- https://docs.google.com/document/d/1KDkGN5guEbZzpn8quSlhJLaM8JhMtmfSrsX0LO05tCs/edit

**Comment by abellotti** [comment]
Here are the different combination of headers cloudigrade handles:


`X-RH-IDENTITY`

- Currently with just the account number:

```
{
  "identity": {
    "account_number": "1234567",
    "user": {
      "is_org_admin": true
    }
  } 
}
```

- With both EBS account number and the org id:

```
{
  "identity": {
    "account_number": "1234567",
    "org_id": "7654321",
    "user": {
      "is_org_admin": true
    }
  } 
}
```

  - With anemic tenants, only the org id will be coming in:
  
```
{
  "identity": {
    "org_id": "7654321",
    "user": {
      "is_org_admin": true
    }
  } 
}
```

For PSK access to cloudigrade, we've added support for the `x-rh-cloudigrade-org-id` header. So the PSK headers we support are:

- x-rh-cloudigrade-psk
- x-rh-cloudigrade-account-number
- x-rh-cloudigrade-org-id

Similar to having the inbound `x-rh-identity` header's account_number, org_id, or just org_id for anemic tenants, the same can be specified to cloudigrade via the psk headers:

```
$ http localhost:8000/api/cloudigrade/v2/instances \
    "x-rh-cloudigrade-psk: <<psk>>" \
    "x-rh-cloudigrade-account-number: 1234567"

$ http localhost:8000/api/cloudigrade/v2/instances \
    "x-rh-cloudigrade-psk: <<psk>>" \
    "x-rh-cloudigrade-org-id: 7654321"
```

**Comment by infinitewarp** [comment]
Independently verified expected handling of all the various supported headers on public v2 API:
 
[![asciicast](https://asciinema.org/a/506464.svg)](https://asciinema.org/a/506464)

Verified org_id on outgoing requests to sources (HTTP and Kafka) via code review of https://github.com/cloudigrade/cloudigrade/pull/1192 (handled in [`generate_sources_headers`](https://github.com/cloudigrade/cloudigrade/blob/b785ff6886e8e7fec21efa54fa075722d43dbaab/cloudigrade/util/redhatcloud/sources.py#L365) and various calls to it).

Looks good to me. Thanks! 🎉
